src/expr/node.h

```diff
@@ -1354,51 +1349,48 @@ NodeTemplate<ref_count>::substitute(TNode node, TNode replacement) const {
   if (node == *this) {
     return replacement;
   }
-  std::unordered_map<TNode, TNode, TNodeHashFunction> cache;
-  return substitute(node, replacement, cache);
+  std::unordered_map<TNode, TNode, TNodeHashFunction> cache{
+      {node, replacement}};
+  return substitute_internal(cache);
 }
 
 template <bool ref_count>
-Node
-NodeTemplate<ref_count>::substitute(TNode node, TNode replacement,
-                                    std::unordered_map<TNode, TNode, TNodeHashFunction>& cache) const {
-  Assert(node != *this);
-
+Node NodeTemplate<ref_count>::substitute_internal(
+    std::unordered_map<TNode, TNode, TNodeHashFunction>& cache) const
+{
+  if (TNode* in_cache = FindOrNull(cache, *this))
+  {
+    return *in_cache;
+  }
+  // TODO(taking): can getNumChildren() == 0 and getMetaKind() == PARAMETERIZED?
   if (getNumChildren() == 0) {
     return *this;
   }
-
-  // in cache?
-  typename std::unordered_map<TNode, TNode, TNodeHashFunction>::const_iterator i = cache.find(*this);
-  if(i != cache.end()) {
-    return (*i).second;
-  }
-
-  // otherwise compute
   NodeBuilder<> nb(getKind());
   if(getMetaKind() == kind::metakind::PARAMETERIZED) {
     // push the operator
-    if(getOperator() == node) {
-      nb << replacement;
-    } else {
-      nb << getOperator().substitute(node, replacement, cache);
+    if (TNode* replacement = FindOrNull(cache, getOperator()))
+    {
+      nb << *replacement;
+    }
+    else
+    {
+      nb << getOperator().substitute_internal(cache);
     }
   }
-  for(const_iterator i = begin(),
-        iend = end();
-      i != iend;
-      ++i) {
-    if(*i == node) {
-      nb << replacement;
-    } else {
-      nb << (*i).substitute(node, replacement, cache);
+  for (const_iterator i = begin(), iend = end(); i != iend; ++i)
+  {
+    if (TNode* replacement = FindOrNull(cache, *i))
+    {
+      nb << *replacement;
+    }
+    else
+    {
+      nb << (*i).substitute_internal(cache);
     }
   }
-
-  // put in cache
   Node n = nb;
-  Assert(node != n);
-  cache[*this] = n;
+  AlwaysAssert(cache.insert(std::make_pair(*this, n)).second);
```

Comment:
We shouldn't have any AlwaysAssert in this function (the performance of substitute is a bottleneck for some applications like quantifiers).  I'd recommend making this Assert.

Reply:
The side effect needs to always happen so 
```
Assert(cache.insert(std::make_pair(*this, n)).second);
```
does not make sense. Would you prefer something like:
```
CVC4_UNUSED const bool inserted = cache.insert(std::make_pair(*this, n)).second;
Assert(inserted);
```

I do not understand the motivation of the comment though. Is the claim that the untaken branch for the assertion has a measurable performance impact? Is it the exceptions? (This can be switched to a CVC4_CHECK. Would that be acceptable?) I am somewhat doubtful this can be distinguished from noise in practice. (Maybe it can on microbenchmarks?)

Reply:
Yes, the original code is a bit confusing, I misinterpreted the fact that what is inside the AlwaysAssert is what is doing the work.

Typically, I expect that what is inside an Assert(...) or AlwaysAssert(...) is side effect free (not sure if there is a strict policy on this).  

The replaced code you give above is what I would expect.

Reply:
I would also prefer keeping Assert(...) and AlwaysAssert(...) side effect free.